Add standard features

sdb9696 opened the pull request:
Some things for discussion here:

- Do we like the `__get_item__` implementation which allows us to have a concise call to a feature:
 `await dev[Feature.state].set_value(True)`
- Do we like the meta class as opposed to an enum to aid the conciseness.
- Do we agree with having a static separate function per standard feature as opposed to a single factory method (I think it's better as they have different signatures depending on the feature)
- Should we consolidate feature testing into a single test module as we'll probably just have one test per feature.
- Should we drop the `lower()` from the feature dictionary keys and keep the key as per the name case?

rytilahti reviewed (commented):
A quick review, couple of answers to the open questions:
> Do we like the __get_item__ implementation which allows us to have a concise call to a feature

While I think the API is actually quite nice, we should separate discussion and implementation into its own PR, and keep this one purely for standard features.

> Do we like the meta class as opposed to an enum to aid the conciseness.

I don't think that makes much sense.

> Do we agree with having a static separate function per standard feature as opposed to a single factory method (I think it's better as they have different signatures depending on the feature)

Different signatures is not a good idea, but it's probably okay to have separate "creators" for the standard types.

> Should we consolidate feature testing into a single test module as we'll probably just have one test per feature.

That sounds reasonable to me.

> Should we drop the lower() from the feature dictionary keys and keep the key as per the name case?

Unsure what's the best approach here.

sdb9696 commented on kasa/feature.py:
```diff
@@ -62,4 +70,38 @@ async def set_value(self, value):
         return await getattr(container, self.attribute_setter)(value)
+
+    @staticmethod
+    def _brightness(device, container=None) -> "Feature":
```
I think this is the best idea, let's go with it

sdb9696 replied:
Done with factory class `StandardFeature`

sdb9696 commented on kasa/feature.py:
```diff
@@ -62,4 +70,38 @@ async def set_value(self, value):
+        return Feature(
+            device=device,
+            container=container,
+            name=Feature.brightness,
```
It was quite verbose following the entire ctor signature for each fixture so I've done it like that for brightness but included a typed `kwargs` solution for rssi and State.  Take a look at the two alternatives and let me know your thoughts.

sdb9696 commented:
> A quick review, couple of answers to the open questions:
> 
> > Do we like the **get_item** implementation which allows us to have a concise call to a feature
> 
> While I think the API is actually quite nice, we should separate discussion and implementation into its own PR, and keep this one purely for standard features.

Removed

> > Do we like the meta class as opposed to an enum to aid the conciseness.

> I don't think that makes much sense.

Removed and gone with constants rather than an enum to keep things simple in the absence of `strenum`.
> 
> > Do we agree with having a static separate function per standard feature as opposed to a single factory method (I think it's better as they have different signatures depending on the feature)
> 
> Different signatures is not a good idea, but it's probably okay to have separate "creators" for the standard types.

Sort of done
> 
> > Should we consolidate feature testing into a single test module as we'll probably just have one test per feature.
> 
> That sounds reasonable to me.

I haven't done that in this PR, can consolidate after.

> > Should we drop the lower() from the feature dictionary keys and keep the key as per the name case?
> 
> Unsure what's the best approach here.

I've dropped it for now so the feature name is the key.

sdb9696 commented on kasa/tests/fixtureinfo.py:
```diff
@@ -15,7 +15,7 @@ class FixtureInfo(NamedTuple):
 
 
-FixtureInfo.__hash__ = lambda x: hash((x.name, x.protocol))  # type: ignore[attr-defined, method-assign]
+FixtureInfo.__hash__ = lambda self: hash((self.name, self.protocol))  # type: ignore[attr-defined, method-assign]
```
This tweak is needed for mypy 1.9.0

rytilahti commented on kasa/__init__.py:
```diff
@@ -58,6 +58,8 @@
     "DeviceType",
     "Feature",
     "FeatureType",
+    "FeatureNames",
+    "StandardFeature",
```
I haven't thought this through nor experimented with it, but I'm wondering if we should just define the standard actions inside `Feature`? Basically leaving us with an API like:
```
await dev.features[Feature.Brightness].set_brightness(123)
await dev.features[Feature.State].turn_on()
```
or even
```
from kasa import Color
await dev.features[Color].set_hsv(1, 1, 1)
```
where `Color` is the name of the interface that a feature implements?

rytilahti commented on kasa/tests/device_fixtures.py:
```diff
@@ -146,6 +146,24 @@ def parametrize_combine(parametrized: List[pytest.MarkDecorator]):
     )
 
 
+def parametrize_subtract(params: pytest.MarkDecorator, subtract: pytest.MarkDecorator):
```
Should be done in a separate PR.

sdb9696 commented:
Closing this for now as things have evolved. Can always go back to it for reference.
